# Analyzing Used Car Listings on eBay Kleinanzeigen

Let's analyze data from used car listings from *eBay Kleinanzeigen*, a classified section from the German eBay website.

The dataset was scraped from the eBay web pages and uploade to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).  The original dataset contains 370,000 used car listings, but we have sampled the data down to 50,000 to ensure this code runs quickly.

This dataset has been edited by [DataQuest](https://dataquest.io) in order to resemble how scraped datasets normally appear in the real world.  The version uploaded to Kaggle has been cleaned.

The dataset contains the following fields:
* dataCrawled - date when the used car ad data was taken.
* name - name of the car.
* seller - indicates if seller is private or a dealer.
* offerType - type of listing.
* price - advertised selling price of the car.
* abtest - indicates if an A/B test is included in the listing.
* vehicleType - vehicle type.
* yearOfRegistration - year which the car was first registered.
* gearbox - transmission type.
* powerPS - power of the car in PS.
* model - car's model.
* kilometer - number of kilometers the car has been driven.
* monthOfRegistration - month of the year which the car was registered.
* fuelType - type of fuel the car uses.
* brand - car's brand.
* notRepairedDamage - any car damage not yet repaired.
* dateCreated - date the eBay listing was created.
* nrOfPictures - number of pictures in the ad.
* postalCode - postal code for the location of the vehicle.
* lastSeenOnline - when the crawler last saw this ad online.

## Examining the Data Set

Let's start by importing the necessary libraries for this project and reading in the dataset and saving it to a variable:

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

Let's examine some data and metadata regarding our used car dataset:

In [2]:
autos.info()
autos.head()
autos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

As we can see, there are 50,000 used car records in the dataset. Of the fields, there are 5 integers and 15 string objects.  The dataset uses more than 7.6 MB.  Seems like most of the strings are in German.  A few of the numeric fields also includes the unit of measure and is stored as a string.  We may need to clean this data to make it usable.

## Convert Column Names to Snake_Case
Python prefers to use column names in snake_case.  Snake_case is the convention of using a single underscore ( _ ) character to separate compound words for a single variable name.

It looks like a few column names use the camel case convention, so we'll convert those into snake case:

In [3]:
autos.rename(columns={"dateCrawled":"date_crawled",
                      "offerType":"offer_type",
                      "vehicleType":"vehicle_type",
                      "yearOfRegistration":"registration_year",
                      "powerPS":"power_ps",
                      "monthOfRegistration":"registration_month",
                      "fuelType":"fuel_type",
                      "notRepairedDamage":"unrepaired_damage",
                      "dateCreated":"ad_created",
                      "nrOfPictures":"num_photos",
                      "postalCode":"postal_code",
                      "lastSeen":"last_seen"},
             inplace=True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

We used the dataframe.rename() function to rename the columns from camelcase to snakecase.

## Further Data Cleaning Tasks

Let's explore our data set some more to see if there is anything else we need to clean up:

In [4]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

### Initial Observations

There are a number of fields where the values are either all the same or have only 2 values.
* seller
* offer_type

The num_photos field seems odd because the mean and standard deviation are zero.  Let's investigate the values in this field further:

In [5]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

It seems like the num_photos field contains only zeroes for every column.  We'll drop this field along with the seller and offer_type fields as they aren't important for our analysis of this dataset.

In [6]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

### Clean Up the Price Field

The price field should be a numeric field, but it is stored as text because it includes the monetary symbol "$" (as seen in in the top value field above).

Let's take a closer look at the price field:

In [7]:
autos["price"].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,200      639
           ... 
$69,999       1
$4,123        1
$24,444       1
$10,556       1
$12,799       1
Name: price, Length: 2357, dtype: int64

Upon further investigation, it looks like the values in the price field includes the dollar sign and comma symbols.  We'll need to remove those symbols and convert this field to a numeric one to make this data useful.

Let's remove the non-numeric characters and convert the price field to a numeric type:

In [8]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].astype(int)
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

### Clean Up the Odometer Field

At first glance, the top value in the odometer field contains a unit of measure ("km").  Let's examine some of the other values in that field as we did with price:

In [9]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

As in the price field, it looks like commas are also included in this field as well.  We'll have to remove the commas and "km" and convert this field to a numeric one.

In [10]:
autos["odometer"] = autos["odometer"].str.replace(",", "")
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].astype(int)
autos["odometer"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer, dtype: int64

To emphasize that the odometer field is in kilometers, we'll rename the column to "odometer_km."

In [11]:
autos.rename(columns={"odometer":"odometer_km"}, 
             inplace=True)
autos.head(n=5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Analysis of Price Field

Now that we have converted the price field to a numeric type, we can do some additional analysis of this field to determine if there is any further cleaning we need to do.

Let's begin by performing an analysis of the prices:

In [12]:
autos["price"].unique().shape

(2357,)

In [13]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [14]:
autos["price"].value_counts().sort_index(ascending=True).head(n=10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

In [15]:
autos["price"].value_counts().sort_index(ascending=False).head(n=10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

It looks like there is a lot of prices that don't make any sense.  For instance, there are 1,421 zero prices and a few prices on the other end of the spectrum that seem to be non-sense (i.e. 99,999,999 , 11,111,111 , 12,345,678).  It wouldn't be unreasonable to pay \\$0 or almost \\$100 million for a used car.

We need set up a criteria for a reasonable price range for a used car (in Germany).  According to the statistical data analysis (the series.describe() method), the median value for used car prices is about \\$2,950.  The 25th percentile value is \\$1,100 and the 75th percentile value is \\$7,200.

It seems reasonable to accept the range of \\$1,000 and \\$10,000 as part of our analysis.  Let's remove the outlier values below:

In [16]:
autos = autos[ autos["price"].between(1000, 10000)]
autos["price"].describe()

count    30567.000000
mean      3991.794321
std       2493.109048
min       1000.000000
25%       1900.000000
50%       3299.000000
75%       5700.000000
max      10000.000000
Name: price, dtype: float64

We went from our initial record count of 50,000 records down to about 30,500 records.  Based on the statistical analysis of our data, the data seems to be more accurate.  The mean about \\$4,000 and the median is about \\$3,300.  It's a lot closer than our inital data, which was mean \\$9,800 and median \\$2,950.

### Analysis of Odometer Field

We'll do the same analysis with the Odometer field as we did with the Price field:

In [17]:
autos["odometer_km"].unique().shape

(13,)

In [18]:
autos["odometer_km"].describe()

count     30567.000000
mean     130805.770929
std       34490.265525
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        346
10000        59
20000       231
30000       293
40000       327
50000       475
60000       589
70000       663
80000       836
90000      1070
100000     1302
125000     3430
150000    20946
Name: odometer_km, dtype: int64

According to the analysis of the odometer data, there doesn't seem to be any values that seem abnormal.  It's probably reasonable that most of the used cars have over 150,000 km (approximately 93,200 miles) and there aren't any cars that have less then 5,000 km (3,100 miles).

## Analysis of Date Fields

The date fields (date_crawled, ad_created, and last_seen) are stored in a string format with the following template:  yyyy-mm-dd hh:mm:ss.  We are only interested in dates and not times and need these dates to be in datetime format so we can perform some analysis on them.  We'll start by removing the time portion of these fields.

In [20]:
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

### date_crawled field

In [21]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025452
2016-03-06    0.013675
2016-03-07    0.035365
2016-03-08    0.033664
2016-03-09    0.032551
2016-03-10    0.034056
2016-03-11    0.033075
2016-03-12    0.038277
2016-03-13    0.015441
2016-03-14    0.037557
2016-03-15    0.033108
2016-03-16    0.029411
2016-03-17    0.030948
2016-03-18    0.012759
2016-03-19    0.033991
2016-03-20    0.038178
2016-03-21    0.037262
2016-03-22    0.032159
2016-03-23    0.031963
2016-03-24    0.028429
2016-03-25    0.030490
2016-03-26    0.033893
2016-03-27    0.031243
2016-03-28    0.035561
2016-03-29    0.033958
2016-03-30    0.033238
2016-03-31    0.031472
2016-04-01    0.033795
2016-04-02    0.035234
2016-04-03    0.038375
2016-04-04    0.036870
2016-04-05    0.013577
2016-04-06    0.003566
2016-04-07    0.001407
Name: date_crawled, dtype: float64

It seems the majority of the data is in a one month period from 3/5/2016 to 4/7/2016.  There seems to be a pretty even distribution of dates.

### ad_created Field

In [22]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-08-10    0.000033
2015-09-09    0.000033
2015-11-10    0.000033
2015-12-30    0.000033
2016-01-07    0.000033
                ...   
2016-04-03    0.038735
2016-04-04    0.037524
2016-04-05    0.012039
2016-04-06    0.003599
2016-04-07    0.001276
Name: ad_created, Length: 66, dtype: float64

There seem to be some really old ads on the eBay site.  Some go far back as a year.

### last_seen Field

In [23]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001210
2016-03-06    0.004188
2016-03-07    0.005300
2016-03-08    0.007099
2016-03-09    0.009880
2016-03-10    0.010665
2016-03-11    0.013119
2016-03-12    0.025060
2016-03-13    0.009324
2016-03-14    0.012824
2016-03-15    0.016325
2016-03-16    0.017208
2016-03-17    0.028495
2016-03-18    0.007917
2016-03-19    0.015965
2016-03-20    0.021134
2016-03-21    0.021003
2016-03-22    0.022017
2016-03-23    0.018680
2016-03-24    0.019891
2016-03-25    0.018844
2016-03-26    0.017175
2016-03-27    0.015147
2016-03-28    0.021134
2016-03-29    0.022083
2016-03-30    0.024340
2016-03-31    0.024046
2016-04-01    0.024405
2016-04-02    0.026074
2016-04-03    0.024929
2016-04-04    0.024602
2016-04-05    0.122812
2016-04-06    0.218667
2016-04-07    0.128439
Name: last_seen, dtype: float64

The majority of the posts seem to be within the week of April 4th through 7th 2016.

### registration_year Field

In [24]:
autos["registration_year"].describe()

count    30567.000000
mean      2004.348906
std         72.163205
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2007.000000
max       9999.000000
Name: registration_year, dtype: float64

In the analysis of registration year, there seems to be some bogus values.  The range is from the year 1000 to the year 9999.  Since cars weren't invented in the year 1000, it's safe to assume we can safely drop that record and the same goes for the year 9999, which is in the future.

For our purposes, we'll only include values in the 1885 to 2016 range.  The first German car was made in 1885 and the 2016 is the latest year according to the analysis of the last_seen, ad_created, and date_crawled fields.

In [25]:
autos = autos[ autos["registration_year"].between(1885, 2016)]
autos["registration_year"].describe()

count    29287.000000
mean      2002.826544
std          6.170955
min       1934.000000
25%       2000.000000
50%       2003.000000
75%       2006.000000
max       2016.000000
Name: registration_year, dtype: float64

Looks like the earliest year for our cleaned up data set is 1934 and goes up to 2016, which seems more reasonable now.

## Analysis of Car Brand

In [26]:
autos["brand"].value_counts()

volkswagen        6272
bmw               3233
opel              3066
mercedes_benz     2885
audi              2292
ford              1904
renault           1316
peugeot            985
fiat               748
smart              612
skoda              563
seat               559
toyota             486
citroen            477
mazda              470
nissan             398
hyundai            328
volvo              270
honda              243
kia                230
sonstige_autos     229
mini               229
mitsubishi         228
alfa_romeo         204
chevrolet          194
suzuki             192
dacia              105
chrysler           104
daihatsu            63
jeep                60
subaru              54
saab                45
land_rover          41
jaguar              38
daewoo              34
trabant             32
porsche             29
rover               26
lada                23
lancia              20
Name: brand, dtype: int64

The description of the data goes here.

In [27]:
car_brands = autos["brand"].value_counts()[:10].index
brand_mean_prices = {}

for car in car_brands:
    brand_mean_prices[car] = autos.loc[ autos["brand"] == car, "price"].mean()

brand_mean_prices

{'volkswagen': 4047.1360012755104,
 'bmw': 4712.516548097742,
 'opel': 3356.864318330072,
 'mercedes_benz': 4331.352512998267,
 'audi': 4596.325916230367,
 'ford': 3485.8550420168067,
 'renault': 2940.159574468085,
 'peugeot': 3405.1015228426395,
 'fiat': 3544.870320855615,
 'smart': 3689.5130718954247}

The top 3 average highest priced cars in this data set are: BMWs at \\$4,700 followed by an Audi at \\$4,600, and Mercedes Benz at \\$4,300.

## Comparing Car Milage to Price

We want to examine if there is a correlation between the milage of the used car to its price.

We first create two dictionaries, one with the car's brand and its price and the other with the car's brand and its milage.  We combine the two dictionaries and create a single dataframe.

Let's begin with converting the dictionary we just built into a data frame:

In [28]:
mean_price_series = pd.Series(brand_mean_prices)
comp_frame = pd.DataFrame(mean_price_series, columns=['mean_price'])
comp_frame

mean_price
volkswagen     4047.136001
bmw            4712.516548
opel           3356.864318
mercedes_benz  4331.352513
audi           4596.325916
ford           3485.855042
renault        2940.159574
peugeot        3405.101523
fiat           3544.870321
smart          3689.513072

Now, let's do the same to the milage column and add it to the dataframe we just created:

In [29]:
brand_mean_milage = {}

for car in car_brands:
    brand_mean_milage[car] = autos.loc[ autos["brand"] == car, "odometer_km"].mean()

mean_milage_series = pd.Series(brand_mean_milage)
comp_frame["mean_milage"] = mean_milage_series

comp_frame.round({"mean_price":2, "mean_milage":0})

mean_price  mean_milage
volkswagen        4047.14     135179.0
bmw               4712.52     142181.0
opel              3356.86     128836.0
mercedes_benz     4331.35     141220.0
audi              4596.33     144282.0
ford              3485.86     126675.0
renault           2940.16     125015.0
peugeot           3405.10     125482.0
fiat              3544.87     110989.0
smart             3689.51      98407.0

There doesn't seem to be any correlation between the price and milage on these cars.